### PR TITLE
Raise exceptions for empty expression stops dictionaries (#9539)

### PR DESCRIPTION
**Pull request by julianrex**

### CODE REVIEW

File: platform/darwin/src/NSExpression+MGLAdditions.mm [hunk 1]
```diff
@@ -1261,9 +1261,15 @@ - (id)mgl_jsonInterpolationExpressionObject {
         NSArray *controlPoints = [self.arguments[curveTypeIndex + 1].collection mgl_jsonExpressionObject];
         [interpolationArray addObjectsFromArray:controlPoints];
     }
+
+    NSDictionary<NSNumber *, NSExpression *> *stops = self.arguments[curveTypeIndex + 2].constantValue;
+
+    if (stops.count == 0) {
```
Comment (fabian-guerra) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: This code assumes that all interpolations must have a stops dictionary. `MGLExpressionInterpolationModeLinear` may not have a stops dictionary.

File: platform/darwin/test/MGLExpressionTests.mm [hunk 2]
```diff
@@ -712,6 +712,31 @@ - (void)testInterpolationExpressionObject {
     }
 }
 
+- (void)testInterpolationExpressionObjectWithEmptyStopsDictionary {
+    {
+        NSDictionary *stops = @{};
+        NSExpression *expression = [NSExpression expressionWithFormat:@"mgl_interpolate:withCurveType:parameters:stops:(x, 'cubic-bezier', { 0.42, 0, 0.58, 1 }, %@)", stops];
+        @try {
```
Comment (fabian-guerra) [inline platform/darwin/test/MGLExpressionTests.mm]: It may be more clear using `XCTAssertThrowsSpecificNamed(expression.mgl_jsonExpressionObject, NSException, NSInvalidArgumentException)` instead.

File: platform/darwin/test/MGLExpressionTests.mm [hunk 3]
```diff
@@ -712,6 +712,31 @@ - (void)testInterpolationExpressionObject {
     }
 }
 
+- (void)testInterpolationExpressionObjectWithEmptyStopsDictionary {
```
Comment (fabian-guerra) [inline platform/darwin/test/MGLExpressionTests.mm]: These tests can be part of `- (void)testInterpolationExpressionObject`

File: platform/ios/Integration Tests/MGLEmptyStopsTests.m [hunk 4]
```diff
@@ -0,0 +1,61 @@
+#import "MGLMapViewIntegrationTest.h"
```
Comment (fabian-guerra) [inline platform/ios/Integration Tests/MGLEmptyStopsTests.m]: There is no need to add an additional test class. All `Expressions` tests are covered in `MGLExpressionTests.mm` and `MGLPredicateTests.mm` respectively.

File: platform/darwin/src/NSExpression+MGLAdditions.mm (hunk 1 above)
Comment (1ec5) [reply platform/darwin/src/NSExpression+MGLAdditions.mm]: Linear interpolation still requires stops. Perhaps you’re thinking of the identity functions of yore.

File: platform/ios/Integration Tests/MGLEmptyStopsTests.m (hunk 4 above)
Comment (1ec5) [reply platform/ios/Integration Tests/MGLEmptyStopsTests.m]: Agreed, this seems like a pretty roundabout way of testing what’s ultimately a minor, pretty self-contained edge case. But if generalized a bit, it could serve as a much more useful integration test of runtime styling.

File: platform/darwin/test/MGLExpressionTests.mm (hunk 2 above)
Comment (julianrex) [reply platform/darwin/test/MGLExpressionTests.mm]: Thanks for introducing me to that one! Saves a ton.

File: platform/ios/Integration Tests/MGLEmptyStopsTests.m (hunk 4 above)
Comment (julianrex) [reply platform/ios/Integration Tests/MGLEmptyStopsTests.m]: I'm in two minds about this. The test was written first as an integration test to try to capture the specific crash (during rendering) mentioned in the original ticket. Recent changes appear to no longer trigger the crash, so this specific test ends up being a somewhat bloated unit-test (and essentially a duplicate of the tests in `MGLExpressionTests.mm`.

I'm tempted to leave it - in case of regression (though perhaps tweaked, and the filename changed) so that it becomes a home for future styling integration tests (as @1ec5 commented).

File: platform/darwin/src/NSExpression+MGLAdditions.mm (hunk 1 above)
Comment (fabian-guerra) [reply platform/darwin/src/NSExpression+MGLAdditions.mm]: You are right. This is not true https://github.com/mapbox/mapbox-gl-native/pull/11723#discussion_r183526278
